### DISCUSSION TITLE
Cannot find name 'BFormGroup' (ts2304) on all components

### DISCUSSION THREAD

**Discussion by mariusa** [cdmoro/bootstrap-vue-3#584] [Q&A]
Using vscode, in `<template>` section it shows all components underlined with red, with error Cannot find name 'BFormGroup' (ts2304) 
![2022-08-25_10-58](https://user-images.githubusercontent.com/1283842/186608859-9f05a4f9-c329-4a1e-b352-be9cee217e32.png)

I see the playground doesn't have this issue, but cannot figure out how to fix it in my project.

Not sure if related, but types are defined in 
```
cat node_modules/bootstrap-vue-3/package.json |grep types
 "types": "./dist/BootstrapVue.d.ts",
    "@types/bootstrap": "^5.2.2",
```

Would you have any hints on where to look?

**Comment by VividLemon** [comment]
I would need to view the dependencies package.json, tsconfig, and main.ts files in your project, in order to be able to solve anything. The "types" file is unrelated

**Reply by mariusa** [reply]
Sure, thanks

dependencies package.json
```js
    "dependencies": {
        "@popperjs/core": "^2.11.5",
        "axios": "0.x.x",
        "bootstrap": "5.x.x",
        "bootstrap-vue-3": "0.x.x",
        "currency.js": "2.x.x",
        "dayjs": "1.x.x",
        "dropzone": "6.0.0-beta.2",
        "mitt": "3.x.x",
        "pluralize": "8.x.x",
        "vite-plugin-pages": "0.x.x",
        "vite-plugin-vue-layouts": "0.x.x",
        "vue": "3.x.x",
        "vue-multiselect": "^3.0.0-alpha.2",
        "vue-plugin-load-script": "2.x.x",
        "vue-router": "4.x.x",
        "vue-router-layout": "0.x.x",
        "vue-runtime-helpers": "1.x.x",
        "vue-sidebar-menu": "5.x.x"
    },
    "devDependencies": {
        "@types/node": "18.x.x",
        "@vitejs/plugin-vue": "3.x.x",
        "@vue/compiler-sfc": "3.x.x",
        "rollup-plugin-visualizer": "^5.5.2",
        "sass": "*",
        "unplugin-auto-import": "0.x.x",
        "unplugin-vue-components": "0.x.x",
        "vite": "3.x.x",
        "vue-template-compiler": "*"
    },
```

tsconfig
```js
{
    "compilerOptions": {
        "allowJs": true,
        "checkJs": true,
        "esModuleInterop": true,
        "importHelpers": true,
        "lib": [
            "esnext",
            "dom",
            "dom.iterable",
            "scripthost"
        ],
        "jsx": "preserve",
        "module": "esnext", // https://github.com/microsoft/TypeScript/issues/24082
        "moduleResolution": "node",
        "noEmit": true,
        "noImplicitAny": false,
        "resolveJsonModule": true,
        "skipLibCheck": true,
        "strict": false,
        "sourceMap": true,
        "target": "esnext",
        "types": [
            "vite/client",
            "vite-plugin-vue-layouts/client",
            "vite-plugin-pages/client",
        ],
        "useUnknownInCatchVariables": false // https://stackoverflow.com/questions/60151181/object-is-of-type-unknown-typescript-generics
    },
    "exclude": [
        "node_modules",
        "**/node_modules/*"
    ],
    "include": [
        "src/**/*.ts",
        "src/**/*.d.ts",
        "src/**/*.tsx",
        "src/**/*.vue"
    ]
}
```

main.ts
```js
import BootstrapVue from 'bootstrap-vue-3' // https://github.com/cdmoro/bootstrap-vue-3  includes bootstrap import
import { BToastPlugin } from 'bootstrap-vue-3' // https://cdmoro.github.io/bootstrap-vue-3/components/Toast.html#toasts-on-demand

import { setupLayouts } from 'virtual:generated-layouts' // https://github.com/JohnCampionJr/vite-plugin-vue-layouts
import generatedRoutes from '~pages' // virtual:generated-pages https://github.com/hannoeru/vite-plugin-pages
const routes = setupLayouts(generatedRoutes)


import { createApp } from 'vue'
import appContainer from './app.vue'

let startApp = async () => {
    const app = createApp(appContainer)
    app.use(BootstrapVue)
    app.use(BToastPlugin)
    app.use(router)

   app.mount('#app')
}

startApp()

```

**Reply by VividLemon** [reply]
How old is this project? Was it created with `pnpm create vite`? (specific package manager doesn't matter). I'd recommend auditing through your dependencies, but the issue may simply be due to the fact that you have a tsconfig, but no TypeScript dependency. To keep things updated and simple, I would just create a new app, then copy over everything besides the package.json dependencies. Then, `pnpm add` all dependencies. (I would exclude ones that seem to be unused, I don't think `vue-template-compiler` is required in Vue 3)

**Reply by mariusa** [reply]
It was duplicated from an older project, don't remember how that was created.

**Reply by VividLemon** [reply]
I can tell, considering @vue/compiler-sfc hasn't been a required dependency since 3.2.13.
Like I said, it may be very beneficial to just merge over into a new project. Alternatively, you could add and remove various dependencies to get it on the same level as a fresh project, but I find that harder.

**Reply by VividLemon** [reply]
These are the only dependencies needed to start a fresh new app
```json
  "dependencies": {
    "vue": "^3.2.37"
  },
  "devDependencies": {
    "@vitejs/plugin-vue": "^3.0.3",
    "typescript": "^4.6.4",
    "vite": "^3.0.7",
    "vue-tsc": "^0.39.5"
  }
```
You could probably remove `@types/node`, `@vue/compiler-sfc`, `vue-template-compiler`. Everything else seems fine, but you'd want to include vue-tsc, typescript, and rollup (not required, but you have an unmet peer dependency). After that, I'd just delete node_modules and install again.

**Reply by mariusa** [reply]
Thanks!

After starting from a fresh project and partially adding code, I've found the issue in using `<any>` inside `<script setup lang="ts">`
Error on `<any>`: `JSX element "any" has no corresponding closing tag`
AboutView.vue
```js

<script setup lang="ts">
let item = $ref({
    id: null,
    detail: <any>{},
})
</script>
<template>
    <div class="about">
        <h1>This is an about page</h1>
        <b-card header="welcome">

            hello
        </b-card>
    </div>

</template>
```

https://github.com/vuejs/eslint-plugin-vue/issues/919

Solution:
```
let item = $ref({
    id: null,
    detail: {} as any,
})
```

**Reply by VividLemon** [reply]
You could alternatively do `let item: Ref<{id: null | any}>`, or `let item = $ref<{id: null | any}>`, or keep it as `detail: {} as any`. But `detail: <any>{},` isn't valid TS. So, it seems you got everything working then?

**Reply by mariusa** [reply]
Yes, appreciate your help!
